### PR TITLE
Apply saved tensor hooks

### LINKED ISSUE #58512: Add hooks to SavedVariables
The goal of this tasks is to provide a convenient hook system for SavedVariables.

The hooks will be a pair of python functions:
- One that goes from the current Tensor to any python object
- A second that will be able to transform that python back into the original Tensor

These operations don't need to be differentiable and any gradient information on the returned Tensor will be overwritten to have proper history.

Implementation details:
- Write a generic c++ class and a python-specialized implementation to hold the hooks (similar to anomaly_mode.h/python_anomaly_mode.h)
- Write a python binding to be able to access the SavedVariable (via pybind like c10::InferenceMode) to be able to add such hook
- Update the SavedVariable to apply such hook (pre-processing when the hook is added and post-processing while unpacking)
  - Autograd property save needs to happen during pre-processing
  - The rewrapping of the returned Tensor should be very similar to the current unpack function
  - The user hooks should run in no-grad mode
  - Inplace tests don't need to happen
- Add the corresponding tests

cc @ezyang @albanD @zou3519 @gqchen @pearu @nikitaved @soulitzer @Lezcano

### PR DESCRIPTION
**Pull request by Varal7**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#60975**

Summary: Fixes #58512
Uses the hooks introduced in #60663: upon registering, the pack hook is called and the returned python object is stored. From then on, whenever we need to unpack it, we will use that python object in combination with the unpack hook.
The packing can be done with gradient tracking disabled as we will add back the correct grad_fn during unpacking.
Inplace operations done by the pack_hook on the original tensor (in the case `leaf || !output`) will be caught if the Saved tensor is used by another op.
Inplace operations done by unpack_hook will unfortunately not be caught. We will add a warning in the docs that follows this PR.

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29466227](https://our.internmc.facebook.com/intern/diff/D29466227)

### CODE REVIEW

**Comment by Varal7** [comment]
@varal7 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29466227).

**Comment by Varal7** [comment]
@varal7 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29466227).

**Comment by Varal7** [comment]
@varal7 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29466227).

**Comment by Varal7** [comment]
@varal7 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29466227).

**Comment by Varal7** [comment]
@varal7 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29466227).

**Comment by Varal7** [comment]
@varal7 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29466227).

**Comment by Varal7** [comment]
@varal7 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29466227).

**Comment by Varal7** [comment]
@varal7 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29466227).

File: torch/csrc/autograd/python_saved_variable_hooks.cpp [hunk 1]
```diff
@@ -9,19 +11,35 @@ namespace torch { namespace autograd {
...
+    data_ = packed.release().ptr(); // steals the reference
+    // pack_hook, obj are decrefed on exit, but we will manually decref data_ when the saved variable is released
+    // pack_hook_ is left as is
   }
 
   at::Tensor PySavedVariableHooks::call_unpack_hook() {
-    TORCH_CHECK_NOT_IMPLEMENTED(false, "Hooks are not implemented yet");
+    py::gil_scoped_acquire acquire;
+    auto unpack_hook = py::reinterpret_borrow<py::function>(unpack_hook_);
+    py::object obj = py::cast<py::object>(data_); // copies the content of data_
```
Comment (albanD) [inline torch/csrc/autograd/python_saved_variable_hooks.cpp:27]: This doesn't actually perform a copy does it? It does get a new reference though?

File: torch/csrc/autograd/python_saved_variable_hooks.cpp [hunk 2]
```diff
@@ -9,19 +11,35 @@ namespace torch { namespace autograd {
     unpack_hook_(unpack_hook.release().ptr()) {}
 
   void PySavedVariableHooks::call_pack_hook(at::Tensor &tensor) {
-    TORCH_CHECK_NOT_IMPLEMENTED(false, "Hooks are not implemented yet");
+    py::gil_scoped_acquire acquire;
+    auto pack_hook = py::reinterpret_borrow<py::function>(pack_hook_); // borrows the reference to the hook
```
Comment (albanD) [inline torch/csrc/autograd/python_saved_variable_hooks.cpp:15]: Do you plan to leave all the ref counting comments?

File: torch/csrc/autograd/saved_variable.cpp [hunk 3]
```diff
@@ -171,6 +209,37 @@ void SavedVariable::register_hooks(std::unique_ptr<SavedVariableHooks>&& hooks)
...
+    } else if (!is_output_) {
+      grad_fn_ = data_.grad_fn();
+    } else {
+      // Current code assumes that the original variable is saved if and only if (is_leaf_ || !is_output)
+      TORCH_INTERNAL_ASSERT(false);
+    }
+    data_ = data_.tensor_data();
+  }
+
+  const bool prev = GradMode::is_enabled();
```
Comment (albanD) [inline torch/csrc/autograd/saved_variable.cpp:238]: You can use `NoGradGuard guard;` to disable autograd as long as that variable exists.

File: torch/csrc/autograd/saved_variable.h [hunk 4]
```diff
@@ -55,6 +53,15 @@ class TORCH_API SavedVariable {
   // a view, see below).
   // The field saved_orignal_ below reflects the two cases: its value is true
   // in the first case and false in the second case.
+  // The value data_.defined() can be false in three cases:
+  // 1. SavedVariable was constructed without a Tensor (the value to save is None), in
+  // that case was_default_constructed_ will be kept at true
+  // 2. The saved variable has been released by calling SavedVariable::reset_data(), typically
+  // during the backward pass
+  // 3. Hooks have been registered. In that case, hooks_ and py_data_ will be defined instead.
```
Comment (albanD) [inline torch/csrc/autograd/saved_variable.h:61]: nit py_data_ doesn't exist anymore

**Review by albanD — COMMENTED** [review]
The high level logic looks good. Just some cleanup to do.

**Comment by Varal7** [comment]
@varal7 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29466227).

File: torch/csrc/autograd/python_saved_variable_hooks.cpp (hunk 1 above)
Comment (Varal7) [reply torch/csrc/autograd/python_saved_variable_hooks.cpp:27]: Yes, sorry my bad, it's only a copy in the sense that we get a new reference.

**Comment by Varal7** [comment]
@varal7 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29466227).

File: torch/csrc/autograd/python_saved_variable_hooks.cpp (hunk 2 above)
Comment (Varal7) [reply torch/csrc/autograd/python_saved_variable_hooks.cpp]: I tried simplifying the ref counting comments

File: torch/csrc/autograd/saved_variable.cpp [hunk 5]
```diff
@@ -57,89 +60,126 @@ SavedVariable::SavedVariable(const Variable& variable, bool is_output, bool is_i
...
+  }
+
   impl::set_version_counter(var, version_counter_);
 
+  // If a Variable is a leaf (no grad_fn saved), and it requires_grad, then we
+  // should have saved the grad accumulator. Even if the Variable is no longer
+  // alive, the accumulator should be kept alive by the references in the
+  // graph.
+  if (requires_grad_ && !var.grad_fn() && grad_accumulator_.expired()) {
+    TORCH_CHECK(false, "No grad accumulator for a saved leaf!");
```
Comment (soulitzer) [inline torch/csrc/autograd/saved_variable.cpp:179]: Should this be TORCH_INTERNAL_ASSERT, or is it possible for user code to reach this point (in which case we could add a test for this?)

**Review by soulitzer — COMMENTED** [review]
Looks mostly good! Just would like to see a couple more cases tested.

File: torch/csrc/autograd/saved_variable.cpp [hunk 6]
```diff
@@ -57,89 +60,126 @@ SavedVariable::SavedVariable(const Variable& variable, bool is_output, bool is_i
...
-                                                       : nullptr;
 
-  if (!saved_original_ && !grad_fn) {
-    TORCH_CHECK(saved_for,"No grad_fn for non-leaf saved variable");
+  auto grad_fn = is_inplace_on_view_ ? weak_grad_fn_.lock()
+                                     : !hooks_ ? saved_original_ ? data_.grad_fn() : nullptr
+                                               : grad_fn_;
+
+  if (!is_leaf_ && !grad_fn) {
+    TORCH_CHECK(saved_for, "No grad_fn for non-leaf saved variable");
```
Comment (soulitzer) [inline torch/csrc/autograd/saved_variable.cpp:117]: Same for this, is this reachable by the user?

File: torch/csrc/autograd/saved_variable.cpp [hunk 7]
```diff
@@ -154,6 +194,9 @@ Variable SavedVariable::unpack(std::shared_ptr<Node> saved_for) const {
 }
 
 void SavedVariable::register_hooks(std::unique_ptr<SavedVariableHooks>&& hooks) {
+  TORCH_CHECK(!hooks_,
```
Comment (soulitzer) [inline torch/csrc/autograd/saved_variable.cpp:197]: Would be good to see a test for this also!

Comment (Varal7) [reply torch/csrc/autograd/saved_variable.cpp:197]: This check was added with the previous PR so it was tested back then (https://github.com/pytorch/pytorch/pull/60663/files#diff-964311d1ab760238d76cb29a0e25e64f2c18cb3dbb076e10d6337c59d6dd718aR4816)

File: torch/csrc/autograd/saved_variable.cpp (hunk 6 above)
Comment (Varal7) [reply torch/csrc/autograd/saved_variable.cpp:117]: This code is not (should not be) reachable from the user.  Will change to TORCH_INTERNAL_CHECK.

File: torch/csrc/autograd/saved_variable.cpp (hunk 5 above)
Comment (Varal7) [reply torch/csrc/autograd/saved_variable.cpp:179]: This code is not (should not be) reachable from the user.  Will change to TORCH_INTERNAL_ASSERT.

File: torch/csrc/autograd/saved_variable.cpp [hunk 8]
```diff
@@ -167,10 +210,28 @@ void SavedVariable::register_hooks(std::unique_ptr<SavedVariableHooks>&& hooks)
         "Calling register_hooks on a saved tensor with value None is forbidden");
     }
   }
-  TORCH_CHECK(!hooks_,
-    "Calling register_hooks on a saved tensor whose hooks have already been set. "
-    "Hint: only one pair of hooks is allowed at a time.");
   hooks_ = std::move(hooks);
+
+  // If we didn't save the original variable, we already have all we need to reconstruct it
+  if (saved_original_) {
```
[the remark this answers is not in the document]
Comment (Varal7) [reply torch/csrc/autograd/saved_variable.cpp:216]: Sounds good, thanks!

File: torch/csrc/autograd/saved_variable.cpp (hunk 7 above)
Comment (soulitzer) [reply torch/csrc/autograd/saved_variable.cpp:197]: Oh whoops, sorry

**Comment by Varal7** [comment]
@soulitzer Thanks for the review. I added tests for three cases:
- Leaf
- Not output
- Output and not leaf (so that `saved_original` is true).
In each case, I check that either the `grad_fn` is correct or the gradient accumulated correctly, depending whether it's a leaf or not.

File: torch/csrc/autograd/saved_variable.cpp [hunk 9]
```diff
@@ -57,89 +60,126 @@ SavedVariable::SavedVariable(const Variable& variable, bool is_output, bool is_i
...
+    var = make_variable(data, requires_grad_);
+  }
+
   impl::set_version_counter(var, version_counter_);
 
+  // If a Variable is a leaf (no grad_fn saved), and it requires_grad, then we
+  // should have saved the grad accumulator. Even if the Variable is no longer
+  // alive, the accumulator should be kept alive by the references in the
+  // graph.
+  if (requires_grad_ && !var.grad_fn() && grad_accumulator_.expired()) {
```
Comment (soulitzer) [inline torch/csrc/autograd/saved_variable.cpp:178]: nit: move this logic into the assert?

**Review by soulitzer — APPROVED** [review]
Thanks for adding the tests!

**Comment by Varal7** [comment]
@varal7 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29466227).
